From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Tue, 2 Sep 2025 13:54:37 -0500
Subject: [PATCH] Fish: Parallel World Ticking API

Original license: MIT
Original project: https://github.com/Biquaternions/Fish

This patch provides an API for performance monitoring plugins like PurpurBars.
A better approach would be to include an event when the world is done ticking,
  which will allow to re-use the RollingAverage logic from Spark and Minecraft internals.
However, since every fork developer will want to have said event into their own namespace,
  it will be virtually impossible to provide a universal API.
With this approach only new methods are added into already existing Bukkit API, and the same
  naming conventions as Bukkit were used, which means there's a bigger chance of this API to
  get standarized.

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1c107711efb8fd41f572f57522c905715e780b0f..496424bee1e37ca66bec786e28620db0a3b9fd52 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3141,4 +3141,11 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Fish start - Parallel World Ticking API
+    @Override
+    public boolean isParallelWorldTickingEnabled() {
+        return org.dreeam.leaf.config.modules.async.SparklyPaperParallelWorldTicking.enabled;
+    }
+    // Fish end - Parallel World Ticking API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index a2b62f5696e911da93bc07b1d438549db4c7b445..8d0e79faf07d5fb38b7aa262e7d1781cdb7e51fd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2113,4 +2113,15 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+
+    // Fish start - Parallel World Ticking API
+    @Override
+    public long @NotNull [] getTickTimes() {
+        return this.world.tickTimes5s.getTimes();
+    }
+    @Override
+    public double getAverageTickTime() {
+        return this.world.tickTimes5s.getAverage();
+    }
+    // Fish end - Parallel World Ticking API
 }
